### PR TITLE
Upgrades the React dependency to v18

### PR DESCRIPTION
**Pull request by jgoz**
### Description

Upgrades the React dependency to v18.

Note that there are several dependent packages whose peer dependency ranges still do not include 18. With the default npm peer dependency mode (strict), this would actually cause `npm install` to fail, so I switched to the legacy peer dependency mode. Keeping strict mode would involve upgrading, or in most cases, switching to alternate dependencies for those that have stopped receiving updates (react-table v6, react-splitter-layout, react-diff-viewer).

It was also not possible to switch to the [new JSX transform](https://legacy.reactjs.org/blog/2020/09/22/introducing-the-new-jsx-transform.html) because react-table has an implicit dependency on the `.createElement` API being used in order to distinguish between class and function components.

<hr>

This PR has:

- [x] been self-reviewed.
- [x] added or updated version, license, or notice information in [licenses.yaml](https://github.com/apache/druid/blob/master/dev/license.md)
- [x] added unit tests or modified existing tests to cover new code paths, ensuring the threshold for [code coverage](https://github.com/apache/druid/blob/master/dev/code-review/code-coverage.md) is met.
- [x] been tested in a test Druid cluster.

### CODE REVIEW

File: web-console/src/singletons/toaster.ts [hunk 1]
```diff
@@ -16,9 +16,16 @@
...
-export const AppToaster = Toaster.create({
-  className: 'recipe-toaster',
-  position: Position.TOP,
-});
+let toaster: ToasterInstance | undefined;
+
+export const AppToaster: { show: ToasterInstance['show'] } = {
+  show: (...args) => {
+    if (!toaster) {
+      toaster = OverlayToaster.create({ className: 'recipe-toaster', position: Position.TOP });
```
Comment (jgoz) [inline web-console/src/singletons/toaster.ts:27]: Making this lazy-init avoids a deprecation notice for "ReactDOM.render", both on initial load and in every unit test.

Comment (vogievetsky) [reply web-console/src/singletons/toaster.ts:27]: That is very insightful! Could you please add that comment (exactly as written) to the code itself?

Maybe also include that the warning is: `react-dom.development.js:86 Warning: ReactDOM.render is no longer supported in React 18. Use createRoot instead. Until you switch to the new API, your app will behave as if it's running React 17. Learn more: https://reactjs.org/link/switch-to-createroot`

File: web-console/package.json [hunk 2]
```diff
@@ -170,5 +171,8 @@
     "webpack-bundle-analyzer": "^4.4.1",
     "webpack-cli": "^4.6.0",
     "webpack-dev-server": "^3.11.2"
+  },
+  "overrides": {
+    "@types/react": "$@types/react"
```
Comment (vogievetsky) [inline web-console/package.json:176]: what does this do and why is it needed?

File: web-console/.npmrc [hunk 3]
```diff
@@ -0,0 +1 @@
+legacy-peer-deps=true
```
Comment (vogievetsky) [inline web-console/.npmrc:1]: I think this file needs the Apache header to pass CI

File: web-console/package.json (hunk 2 above)
Comment (jgoz) [reply web-console/package.json:176]: Some dependencies include `@types/react` as a dependency but pinned to specific versions. This forces those transitive dependency versions to resolve to the version specified in our package file.

**Pull request merged**